[decisions] keyset types (#4717)
ElektraInitiative/libelektra: doc/decisions/0_drafts/keyset_types.md

# Types of `KeySet`s

## Problem

A `KeySet` can be used in different ways.
Among these are at least:

- As originally intended, a set of configuration data
- Metadata of a `Key`

> - General associative array data structure

Reviewer: This would be perfect to have in use cases.

Author: https://github.com/ElektraInitiative/libelektra/blob/3e9036fc91533cc03dfbb6c3fc0a01cf45f12788/doc/usecases/core/README.md#L5-L12

These types have slightly different requirements, and therefore may need to be treated differently.

## Constraints

> - API duplication just to separate types should be avoided

Reviewer: I agree about the goal but I don't think its a constraint.

```suggestion
- API must be minimal
```

Author: You could argue the API would still be minimal, since you cannot reduce it without loosing type safety or functionality.

Author: Maybe we make it an assumption:

"Introducing a type safe API with different C-types for each type of `KeySet`, is not worth the necessary API duplication"

- API must be safe to use for each of the different types

## Assumptions

> - The types are not different enough to warrant the overhead of entirely different APIs and data structures.

Reviewer: Assumption is okay :+1:

## Considered Alternatives

### No restrictions

Leave `KeySet` entirely unrestricted, so it can be used for each use case.

Sometimes there can be odd edge cases.

> For example, what should a storage plugin do, if it receives a `KeySet` without both configuration data and metadata.

Reviewer: Put this up to the problem.

Author: Just noticed a typo here:

```suggestion
Sometimes there can be odd edge cases.
For example, what should a storage plugin do, if it receives a `KeySet` with both configuration data and metadata.
```

I'll add it to the problem as well.

### Restriction based on first `Key`

To effectively create different kinds of `KeySet`s, we restrict which `Key`s can be inserted into a `KeySet` based on the first `Key` in a `KeySet`.

> Any `Key` can be inserted into an empty `KeySet`.

Reviewer: ```suggestion
Any `Key` except cascading keys can be inserted into an empty `KeySet`.
```

Only `Key`s matching the type of a `KeySet` can be inserted into a non-empty `KeySet`.
The type of a `KeySet` is determined by the first `Key`, i.e. `ksAtCursor(ks, 0)`.
Transitively, this means the type of a `KeySet` is fixed when the first `Key` is inserted and when a `KeySet` is cleared, it becomes "type-less" again.

The different types of `KeySet` will be:

1. Metadata:
   Can only contain `Key`s with namespace `KEY_NS_META`.
2. General data:

>    Can only contain `Key`s with namespace `KEY_NS_DATA`.

Reviewer: KEY_NS_DATA should be explained somewhere, probably best in its use case.

Author: I think I'll add a new decision to this PR that introduces `KEY_NS_DATA`. It should be a very short and simple decision and under the constraint of this decision there is only one alternative: "Don't use `KeySet` for general data".

3. Config data:
   Can only contain a mix of `Key`s with namespace `KEY_NS_SPEC`, `KEY_NS_PROC`, `KEY_NS_DIR`, `KEY_NS_USER`, `KEY_NS_SYSTEM` and `KEY_NS_DEFAULT`.

Importantly, `Key`s with namespace `KEY_NS_CASCADING` can never be inserted into a `KeySet`.
This is a restriction that simplifies the logic for cascading lookups.

> This can now be achieved with the new `KEY_NS_DATA` namespace.

Reviewer: Please clarify what "This" is.

Author: ```suggestion
Instead the `KEY_NS_DATA` namespace should be used.
```

Cascading lookups always try the namespaces allowed in the type of `KeySet` in order, but never try `KEY_NS_SPEC`.
This means cascading lookups work for all types of `KeySet` the way a user would expect.

### Different structs and APIs

Extract the implementation of `KeySet` into an internal API (and possibly data structure).
Expose a different type and API for each of the use cases.

This creates unnecessary overhead and at least some code duplication, even if it is minimized by extracting the implementation.

> 

Reviewer: ```suggestion

### KeySets also have Namespaces

Add a `ksSetNamespace` API which allows to change the namespace of a KeySet.
This requires an alias `KEY_NS_CONFIG` to the namespaces `KEY_NS_SPEC`, `KEY_NS_PROC`, `KEY_NS_DIR`, `KEY_NS_USER`, `KEY_NS_SYSTEM` and `KEY_NS_DEFAULT`.

Only keys of the namespace set to `ksSetNamespace` can be added.
`ksSetNamespace` fails if called after keys were added.

This alternative conflicts with the constraint of minimal API.
```

Author: It shouldn't be `KEY_NS_CONFIG`. There should be a separate enum

```c
enum KeySetType {
    KS_TYPE_NONE,
    KS_TYPE_CONFIG,
    KS_TYPE_META,
    KS_TYPE_DATA
};
```

## Decision

**Suggestion:** Go with "Restriction based on first `Key`"

## Rationale

Some kind of separation is necessary to avoid issues with e.g. `KeySet * keyMeta(Key * key)`

> This way we avoid API/code duplication and still allow slightly different handling of the different types.

Reviewer: Better avoid Rationale now, as it is unclear to what it refers.

Please describe a solution for metadata above in the "Considered Alternatives". We obviously don't want that someone removes all keys and mutates a meta-data keyset to some other keyset later on.

Author: > We obviously don't want that someone removes all keys and mutates a meta-data keyset to some other keyset later on.

Hm, yeah that's an issue. How about we add a flag

```
bool forceMeta : 1; /**< if set the KeySet can only contain meta:/ keys */
```
